src/machine/eval.rs

```diff
@@ -123,46 +123,42 @@ impl<W: Write> Machine<W> {
         match prim {
             Primitive::Print => {
                 writeln!(self.env.stdout, "{}", arg).expect("Primitive print failed");
-                (true, Term::Lit(Literal::Unit))
+                (true, Term::Lit(0))
             }
         }
     }
 }
 
-fn native_bin_op(op: BinOp, l1: Literal, l2: Literal) -> Literal {
+fn native_bin_op(op: BinOp, n1: i64, n2: i64) -> i64 {
     use BinOp::*;
-    use Literal::*;
 
-    match (op, l1, l2) {
-        (Add, Number(n1), Number(n2)) => (n1 + n2).into(),
-        (Sub, Number(n1), Number(n2)) => (n1 - n2).into(),
-        (Mul, Number(n1), Number(n2)) => (n1 * n2).into(),
-        (Div, Number(n1), Number(n2)) => (n1 / n2).into(),
-        (Rem, Number(n1), Number(n2)) => (n1 % n2).into(),
-        (Lt, Number(n1), Number(n2)) => (n1 < n2).into(),
-        (Lte, Number(n1), Number(n2)) => (n1 <= n2).into(),
-        (Gt, Number(n1), Number(n2)) => (n1 > n2).into(),
-        (Gte, Number(n1), Number(n2)) => (n1 >= n2).into(),
-        (Eq, l1, l2) => (l1 == l2).into(),
-        (Neq, l1, l2) => (l1 != l2).into(),
-        (And, Bool(b1), Bool(b2)) => (b1 && b2).into(),
-        (Or, Bool(b1), Bool(b2)) => (b1 || b2).into(),
-        (BitAnd, Number(n1), Number(n2)) => (n1 & n2).into(),
-        (BitOr, Number(n1), Number(n2)) => (n1 | n2).into(),
-        (BitXor, Number(n1), Number(n2)) => (n1 ^ n2).into(),
-        (Shr, Number(n1), Number(n2)) => (n1 >> n2).into(),
-        (Shl, Number(n1), Number(n2)) => (n1 << n2).into(),
-        (op, l1, l2) => panic!("Unexpected operation `{} {} {}`", l1, op, l2),
+    match op {
+        Add => n1 + n2,
+        Sub => n1 - n2,
+        Mul => n1 * n2,
+        Div => n1 / n2,
+        Rem => n1 % n2,
+        Lt => (n1 < n2).into(),
+        Lte => (n1 <= n2).into(),
+        Gt => (n1 > n2).into(),
+        Gte => (n1 >= n2).into(),
+        Eq => (n1 == n2).into(),
+        Neq => (n1 != n2).into(),
+        And => (n1 == 1 && n2 == 1).into(),
```

Comment:
assuming the traditional layout of booleans this could be just `n1 & n2`. Something similar happens for `Or`

Reply:
Merged `And, BitAnd` and `Or, BitOr`